### ISSUE TITLE
关于Amazon-2018-Scientific数据集统计数据问题

### ISSUE THREAD

**Issue by AlwaysFHao** [RUCAIBox/UniSRec#25]
您好，感谢你们优秀的工作UniSRec，这篇工作给到我很大的启发，但我在对齐实验结果时出现了一些问题。我尝试直接通过Recbole框架在Amazon-Scientific数据集上运行SASRec，数据集源文件和数据处理都是遵循Recbole官方提供的内容（也进行了5-core过滤），但是各项指标都远高于你们论文中展示的SASRec的性能，甚至超过UniSRec。通过对比数据集统计信息，我发现Recbole官方的数据集统计信息与你们论文中的统计信息无法对齐，具体数据如下：#users=10716, #items=5014, #inters=74837。我也尝试自己处理Scientific数据集，通过5-core过滤后也和Recbole的信息是能对上的，但与你们论文中的数据对不上。另外我还观察到UniSRec在论文中提到是遵循Recbole范式的，但现在数据集统计数据出现了偏差，我想询问一下出现这种数据集偏差问题的原因？我希望能够对齐你们的实验结果，万分感谢！

虽然不知道具体差异在哪，但我觉得可能是5-core过滤实现的区别。以下是我实现的5-core过滤与UniSRec的5-core过滤的区别，我的实现参照了[MMSRec](https://github.com/kz-song/MMSRec)在Amazon2014数据集的处理方法。
![2E628B576383B1DD3B8C0749EC36645D](https://github.com/user-attachments/assets/ac7b4c7c-de56-4dfe-907e-d22095746ebb)

**Comment by hyp1231** [comment]
您好，请问您处理的数据是和 MMSRec 一样的 Amazon 2014 版本的数据吗？

Amazon Reviews 目前有 4 个版本，分别是 2013、2014、2018 和 2023 版本。在 UniSRec 的 paper 第五页，我们提到使用的是
> Jianmo Ni, Jiacheng Li, and Julian J. McAuley. 2019. Justifying Recommendations using Distantly-Labeled Reviews and Fine-Grained Aspects. In EMNLP.

里使用的 Amazon Reviews 数据集，即 Amazon 2018；在我们的 GitHub Repo 的数据预处理部分[https://github.com/RUCAIBox/UniSRec/tree/master/dataset#amazon-2018](https://github.com/RUCAIBox/UniSRec/tree/master/dataset#amazon-2018)我们也标注了使用的是 Amazon 2018 版本数据集。所以我推测统计信息不一致可能是因为我们使用了不同版本的数据集？

**Comment by AlwaysFHao** [comment]
> 您好，请问您处理的数据是和 MMSRec 一样的 Amazon 2014 版本的数据吗？
> 
> Amazon Reviews 目前有 4 个版本，分别是 2013、2014、2018 和 2023 版本。在 UniSRec 的 paper 第五页，我们提到使用的是
> 
> > Jianmo Ni, Jiacheng Li, and Julian J. McAuley. 2019. Justifying Recommendations using Distantly-Labeled Reviews and Fine-Grained Aspects. In EMNLP.
> 
> 里使用的 Amazon Reviews 数据集，即 Amazon 2018；在我们的 GitHub Repo 的数据预处理部分https://github.com/RUCAIBox/UniSRec/tree/master/dataset#amazon-2018我们也标注了使用的是 Amazon 2018 版本数据集。所以我推测统计信息不一致可能是因为我们使用了不同版本的数据集？

感谢您的回复！我知道Amazon数据集有多个版本差异，我在处理数据时下载的是2018版本的数据集，而且Scientific数据集没有2014版本。不过我已经发现了出现偏差的问题所在，其实UniSRec与MMSRec的处理结果是相同的，我为了复现RecBole对Amazon数据集的处理结果，所以inters的读取使用了list存放保证顺序，而UnisRec和MMSRec用了set。按照常理来说inters不会有重复，2014数据集确实如此，但是2018数据集inters存在重复条目，这就导致item 5-core过滤时会有部分实际4次交互的item被统计成5次。我将inters的读取使用set存放去重，就可以得到与你们一致的处理结果了。

但是RecBole框架的处理貌似使用的还是list，并未考虑重复的情况，因为我使用recbole官方提供的2018数据集得到的统计结果是与我使用list存放inters一致的，我猜测recbole为了保证每次生成item的id映射顺序，所以使用了list（实际上不同的id顺序确实会对模型性能产生较大影响，特别是涉及ID Embedding学习时）。

另外我还存在一个疑问，请问你们预训练的5个数据集是使用什么策略进行合并的呢？因为我观察到你们提供的预训练数据集是合并好的，但我对合并的细节有点疑惑，比如各个数据集的item id怎么映射等处理细节。期待您的回复，万分感谢！

**Comment by hyp1231** [comment]
哇感谢指出这个不一致的地方，感觉很难找出来，谢谢你的细心！

Amazon Reviews 这个数据集中的 interactions 都是 user 对 item 评论，按道理来说不太应该出现重复的 user-item 交互，所以感觉用 set 也算合理；如果考虑一般的 sequential rec，可能默认用 list 会更好一些，因为当 interaction 是点击/购买等行为时，行为序列中出现重复的 item 也是很正常的。不过总的来说，不同的数据处理方式可能不会特别影响实验结论，只要保证比较不同模型时使用一种固定的处理方式即可。

关于预训练数据处理，主要的代码在 [to_pretrain_atomic_files.py](https://github.com/RUCAIBox/UniSRec/blob/master/dataset/preprocessing/to_pretrain_atomic_files.py) 中，可以参考数据处理 README 中的[脚本](https://github.com/RUCAIBox/UniSRec/blob/master/dataset/README.md#3-process-pretrain-datasets)。

其中关于五个数据集合并的代码如下：

https://github.com/RUCAIBox/UniSRec/blob/05aa5cba2809112c32808f70d16abc61c05c6538/dataset/preprocessing/to_pretrain_atomic_files.py#L19-L53

在合并时，我们要保证离线的数据集文件中，不同 domain/category 的 item IDs 不会互相冲突，这样 RecBole 加载后会自动进行 remapping，把字符串格式的 item IDs 映射成整数 IDs。我们处理好的预训练数据中，每个字符串格式的 item ID 是 `{领域编号}-{商品编号}` 这样表示的。关于合并策略，我们只是把五个数据集的 item sequences 拼在一起。训练过程中，dataloader 会自动进行 shuffle。

**Comment by AlwaysFHao** [comment]
> 哇感谢指出这个不一致的地方，感觉很难找出来，谢谢你的细心！
> 
> Amazon Reviews 这个数据集中的 interactions 都是 user 对 item 评论，按道理来说不太应该出现重复的 user-item 交互，所以感觉用 set 也算合理；如果考虑一般的 sequential rec，可能默认用 list 会更好一些，因为当 interaction 是点击/购买等行为时，行为序列中出现重复的 item 也是很正常的。不过总的来说，不同的数据处理方式可能不会特别影响实验结论，只要保证比较不同模型时使用一种固定的处理方式即可。
> 
> 关于预训练数据处理，主要的代码在 [to_pretrain_atomic_files.py](https://github.com/RUCAIBox/UniSRec/blob/master/dataset/preprocessing/to_pretrain_atomic_files.py) 中，可以参考数据处理 README 中的[脚本](https://github.com/RUCAIBox/UniSRec/blob/master/dataset/README.md#3-process-pretrain-datasets)。
> 
> 其中关于五个数据集合并的代码如下：
> 
> https://github.com/RUCAIBox/UniSRec/blob/05aa5cba2809112c32808f70d16abc61c05c6538/dataset/preprocessing/to_pretrain_atomic_files.py#L19-L53
> 
> 在合并时，我们要保证离线的数据集文件中，不同 domain/category 的 item IDs 不会互相冲突，这样 RecBole 加载后会自动进行 remapping，把字符串格式的 item IDs 映射成整数 IDs。我们处理好的预训练数据中，每个字符串格式的 item ID 是 `{领域编号}-{商品编号}` 这样表示的。关于合并策略，我们只是把五个数据集的 item sequences 拼在一起。训练过程中，dataloader 会自动进行 shuffle。

我也考虑过重复点击的问题，但是inters里重复的条目就连时间戳都是重复的，我怀疑是Amazon2018原始数据集处理的问题？所以确实应该是set更合理，就是不清楚为何rebole用的list，没有去重。

感谢你对预训练数据集处理部分的回复，我大概明白了处理流程。

另外我还想询问一下，Recbole运行baseline时，我该如何直接使用unisrec预处理好的数据集文件呢？因为我的理解里，recbole运行Sequential Recommendation模型时，输入的数据集还是.inter文件，然后运行时再做序列化处理，请问我该如何将处理好的.train等文件作为输入呢？Recbole源码对这部分的封装让我有点摸不着头脑。

**Comment by hyp1231** [comment]
如果希望跑一般的 sequential rec baseline，可以直接运行这个仓库里的 `run_baseline.py`，应该可以直接运行大部分 RecBole 内置的模型如 `SASRec`。

如果希望使用原生 RecBole，比较关键的配置是：

```yaml
benchmark_filename: [train, valid, test]
alias_of_item_id: [item_id_list]
load_col:
  inter: [user_id, item_id_list, item_id]
train_neg_sample_args: ~
```

这样可以从 "benchmark" 数据集中加载已经预处理好的交互序列，而无需 RecBole 处理，保证数据处理的完全一致。

**Comment by AlwaysFHao** [comment]
> 如果希望跑一般的 sequential rec baseline，可以直接运行这个仓库里的 `run_baseline.py`，应该可以直接运行大部分 RecBole 内置的模型如 `SASRec`。
> 
> 如果希望使用原生 RecBole，比较关键的配置是：
> 
> ```yaml
> benchmark_filename: [train, valid, test]
> alias_of_item_id: [item_id_list]
> load_col:
>   inter: [user_id, item_id_list, item_id]
> train_neg_sample_args: ~
> ```
> 
> 这样可以从 "benchmark" 数据集中加载已经预处理好的交互序列，而无需 RecBole 处理，保证数据处理的完全一致。

好的，感谢您的详细讲解！祝一切顺利！
